Docstring-based `add_py_function_tool()`

grf53 opened the pull request:
## Ticket
#38

## Description
Add feature to put python callable to the `add_py_function_tool()` without extra tool description JSON document, the function's docstring is parsed into the tool description(JSON Schema format).

Now it only supports google-style docstring just like the transformers.

### A Breaking Change
The parameter order of the `add_py_function_tool()` is changed
from: `def add_py_function_tool(self, desc: dict, f: Callable[..., Any]):`
to: `def add_py_function_tool(self, f: Callable[..., Any], desc: Optional[dict] = None):`
in order to make the tool description optional.

This had better to be merged quickly not to confuse users.

jhlee525 reviewed (approved):
👍 

Note that we have to update our reference API (until the next release.)

grf53 commented:
> Note that we have to update our reference API (until the next release.)

Just checked that pdoc generates document with modified APIs

grf53 commented:
Changed the validation logic for tool description in `add_py_function_tool()`
1. When the tool description(`desc`) is explicitly given as parameter and the validation is failed for that description, it just raises the `ValidationError` instead of trying to use parsed description.
2. To make the code clearer and easier to see where the errors raised, I split up the parts that call `get_json_schema()` and `ToolDescription.model_validate()`.
3. (Added docstring describing possible errors from the function.)
